Vending Machine improvements

DeityLink opened the pull request:
- Vending machines articles can now be given a "category". Items of the same category will show together in the vending machine's menu.
- Contraband and premium items will still appear at the bottom of the window, in their own category.
- Added categories for most seeds. Categories are made from a culinary standpoint, thus tomatoes and pumpkins are among vegetables.
- YOU CAN FINALLY ADD SEEDS TO THE MEGASEED VENDOR, WOO!! You can load a bunch of them at once by using a plant bag.

![update](https://cloud.githubusercontent.com/assets/7573912/8123659/2c658820-10ce-11e5-8a2f-6e12cf15a111.png)

This should have been a feature since long ago.

PJB3005 commented:
It'd be nice if you'd make them use nanoUI or atleast use browser datums.

DeityLink commented:
>NanoUI on vending machines
over my dead body.

ComicIronic commented on code/game/machinery/vending.dm:
```diff
@@ -328,6 +334,27 @@
+			if(bag.contents.len > 0)
+				user << "<span class='notice'>Some items are refused.</span>"
+	else
+		for(var/typepath in allowed_inputs)
```
if(is_type_in_list(W, allowed_inputs))

Also return 1 after this

ComicIronic commented on code/game/machinery/vending.dm:
```diff
@@ -479,6 +518,52 @@
+		for(var/cat_name in category_names)
+			dat += {"<B>&nbsp;&nbsp;[cat_name]</B>:<br>"}
+			for (var/datum/data/vending_product/R in categories[cat_name])
+				dat += {"<FONT color = '[R.display_color]'><B>[R.product_name]</B>:
```
Lots and lots of copypaste.

I'd suggest making this a vending_product proc.

ComicIronic commented on code/game/atoms.dm:
```diff
@@ -48,6 +48,8 @@ var/global/list/ghdel_profiling = list()
 	//var/harm_label_icon_state //Makes sense to have this, but I can't sprite. May be added later.
 	var/list/last_beamchecks // timings for beam checks.
 
+	var/vending_cat = null// subcategory for vending machines.
```
This is really not how I would implement this.
Ideally, products should be a list of new vending datums which you can pass an arg like category and price to.

DeityLink replied:
When we implement custom vending machines it'll be easy to code an override of that var, so players can set up their own categories and prices. Think of this as a "default" var, for items that you want automatically added to the machine, like for instance, seeds.

ComicIronic replied:
At least move this to movable atoms.
There's no sense having it bloating turf memory.

Though I can't think of a vending machine that gives out anything other than items.

DeityLink replied:
eh, you're right, not sure why https://github.com/d3athrow/vgstation13/pull/4935/files#diff-be8fc4ef536d83a7c390019dd59d14efR228 was a var/atom when it could have been a var/obj/item. I'm changing that.

PJB3005 commented:
@DeityLink #4226

Both.

DeityLink commented:
Well I approve of that.

DeityLink commented:
aight, should be good now

FudgePucker commented:
What about if I put contraband seeds in?
What about geneticly altered contraband like death nettles, poison apples and death shrooms?

DeityLink commented:
contraband seeds will be stored among regular seeds.

the Seed Extractor is already designed to store genetically altered seeds.

DeityLink commented:
fixed a conflict caused by whoever added glass fuel cans or whatever to engi vends.

ririchiyo commented on code/game/machinery/vending.dm:
```diff
@@ -396,6 +419,20 @@
+		var/idx=GetProductIndex(P)
+		dat += " <a href='byond://?src=\ref[src];vend=[idx];cat=[P.category]'>(Vend)</A>"
+	else
+		dat += " <font color = 'red'>SOLD OUT</font>"
```
change this to a span tag.

ririchiyo commented on code/game/machinery/vending.dm:
```diff
@@ -464,20 +503,40 @@
+				categories["default"] += R
+
+		for (var/datum/data/vending_product/R in categories["default"])
+			dat += GetProductLine(R)
```
don't do this convert this stuff to docstrings {} instead of adding dat += multiple times strings are slow in BYOND.

ririchiyo commented:
This has some trailing commas along with some spacing issues that make it really confusing to look at. And please do at least convert this to utilize the browser datum system.

ComicIronic commented on code/game/machinery/vending.dm:
```diff
@@ -221,7 +225,7 @@
 		if(isnull(amount)) amount = 1
 
-		var/atom/temp = new typepath(null)
+		var/obj/item/temp = new typepath(null)
```
There's no need to do a new here.
You can just cast the typepath and use initial() for the vars.

ComicIronic commented:
As cool as a feature this is, it's going to be the straw that broke the camel's back in terms of how back the starting code for vending machines is.

They need to move to using vending datums right off the bat, instead of generating them from a type list. The vending datum's New could then be used to set price, category, subcategory, etc, and I don't want to add more workload to that eventual and very necessary change.

DrCelt commented:
@ComicIronic
That's what I was planning to do with vending machines in my PR a while back which I think I could get onto doing again once this one merges. I was also planning to figure out how to make it nanoUI but that's a lot of effort too.

DeityLink commented:
I'll see about that once E3 is over.

DeityLink commented:
Alright, fixed issues, fixed conflicts, aaaaand added categories for booze:

![dairy](https://cloud.githubusercontent.com/assets/7573912/8216342/3f892c6a-1536-11e5-9d6b-89c57256561f.png)

you cannot re-add bottles to it though.

PJB3005 commented:
still no browser datums...

DeityLink commented:
that's beyond the scope of this PR. I'm not going to revamp vending machine browsers in there.

PJB3005 commented:
it's literally a 2 line change.

DeityLink commented:
Well I really don't want vending machines menus to look like that so no. I'm not adding it in this PR, period.
